cudamapper/src/index_cache.cu

```diff
@@ -41,84 +41,164 @@ IndexCacheHost::IndexCacheHost(const bool reuse_data,
         {
             // check if the same index already exists in the other cache
             auto existing_cache = cache_to_check.find(descriptor_of_index_to_cache);
             if (existing_cache != cache_to_check.end())
             {
                 index_copy = existing_cache->second;
+                if (keep_on_device)
+                {
+                    auto existing_device_cache = temp_device_cache_to_check.find(descriptor_of_index_to_cache);
+                    if (existing_device_cache != temp_device_cache_to_check.end())
+                    {
+                        index_on_device = existing_device_cache->second;
+                    }
+                    else
+                    {
+                        index_on_device = index_copy->copy_index_to_device(allocator_, cuda_stream_);
+                    }
+                }
             }
         }
 
         if (nullptr == index_copy)
         {
-            // create index
-            auto index = claragenomics::cudamapper::Index::create_index(allocator_,
-                                                                        *parser,
-                                                                        descriptor_of_index_to_cache.first_read(),
-                                                                        descriptor_of_index_to_cache.first_read() + descriptor_of_index_to_cache.number_of_reads(),
-                                                                        kmer_size_,
-                                                                        window_size_,
-                                                                        hash_representations_,
-                                                                        filtering_parameter_,
-                                                                        cuda_stream_);
-            // copy it to host memory
-            index_copy = claragenomics::cudamapper::IndexHostCopy::create_cache(*index,
-                                                                                descriptor_of_index_to_cache.first_read(),
-                                                                                kmer_size_,
-                                                                                window_size_,
-                                                                                cuda_stream_);
+            // check if this index is already cached in this cache
+            auto existing_cache = cache_to_edit.find(descriptor_of_index_to_cache);
+            if (existing_cache != cache_to_edit.end())
+            {
+                // index already cached
+                index_copy = existing_cache->second;
+                if (keep_on_device)
+                {
+                    index_on_device = index_copy->copy_index_to_device(allocator_, cuda_stream_);
+                }
+            }
+            else
+            {
+                // create index
```

Comment:
Are duplicate entries in `descriptors_of_indices_to_cache` expected?
This function works for duplicate entries, but it will redo the index generation and then overwrite the previous cached index. If there are frequent duplicate entries, you might want to check the fresh cache, too.

Reply:
No, every `IndexDescriptor` in `descriptors_of_indices_to_cache` is expected to be different.
I could add an assertion to check for that, but as the program would work correctly in that case as well I don't think it's necessary

Reply:
I agree, I don't think an assert is necessary here. Just wanted to check if this is a common case.